vertical-pod-autoscaler/pkg/recommender/input/cluster_feeder.go

```diff
@@ -192,13 +202,61 @@ type clusterStateFeeder struct {
 	memorySaveMode      bool
 	controllerFetcher   controllerfetcher.ControllerFetcher
 	recommenderName     string
+	historySource       HistorySource
+	promHistoryConfig   history.PrometheusHistoryProviderConfig
+}
+
+func (feeder *clusterStateFeeder) Init() error {
+	switch feeder.historySource {
+	case Checkpoints:
+		klog.Infof("Using checkpoints as a history provider")
+		feeder.initFromCheckpoints()
+	case Prometheus:
+		klog.Infof("Using prometheus as a history provider")
+
+		provider, promInitErr := history.NewPrometheusHistoryProvider(feeder.promHistoryConfig)
+		if promInitErr != nil {
+			klog.Errorf("Could not initialize history provider")
+			return promInitErr
+		}
+
+		historyInitErr := feeder.initFromHistoryProvider(provider)
+		if historyInitErr != nil {
+			klog.Errorf("Failed to load prometheus history")
+			return historyInitErr
+		}
+	case None:
+		klog.Infof("Running without a history provider")
+	default:
+		klog.Errorf("Wrong storage provider option")
+		return fmt.Errorf("storage provider option is not set. Supported values: prometheus, none, checkpoint")
+	}
+	return nil
 }
 
-func (feeder *clusterStateFeeder) InitFromHistoryProvider(historyProvider history.HistoryProvider) {
+func GetHistorySourceFromArg(historySource string) (HistorySource, error) {
+	switch historySource {
+	case "checkpoint":
+		return Checkpoints, nil
+	case "prometheus":
+		return Prometheus, nil
+	case "none":
+		return None, nil
+	default:
+		klog.Error("Storage option '%s' is not supported. Supported values: prometheus, none, checkpoint", historySource)
+		return Undefined, fmt.Errorf("storage option '%s' is not supported. Supported values: prometheus, none, checkpoint", historySource)
+	}
+}
+
+func (feeder *clusterStateFeeder) initFromHistoryProvider(historyProvider history.HistoryProvider) (historyInitError error) {
+	historyInitError = nil
 	klog.V(3).Info("Initializing VPA from history provider")
 	clusterHistory, err := historyProvider.GetClusterHistory()
 	if err != nil {
-		klog.Errorf("Cannot get cluster history: %v", err)
+		historyInitError = err
+	}
+	if len(clusterHistory) == 0 {
+		klog.Warningf("history provider returned no pods")
```

Comment:
This potentially could also be an error/restart, but not sure if any valid cases are possible, so leaving with a warning for now.

As for the error cases, at the moment `--metric-for-pod-labels=up{job=\"kubernetes-pods\"}`, if prom is not configured to have a given job label, the query will not return any results, so history will not be used (and users won't be notified about the misconfiguration)

Reply:
This case does not seem to be treated differently (e.g., issue warning logs) before this PR. Is anything changed in this PR that make this case special? Or you just hope to add a log for this case?

Reply:
Yes, this is behaviour preexisting before this PR. (also it can be a valid behaviour, say when a user specified the metric filter to target a subset of pods, but no such pods existed in the required history window).

Adding warning here just in case when this behaviour is not intentional, so it is easier to uncover a misconfiguration, for example see https://github.com/kubernetes/autoscaler/issues/6221 and https://github.com/kubernetes/autoscaler/issues/5031#issuecomment-1450583325.